File: algo-c-to-_/src/svgplot.py
#
#	from src/svgplot.c
#
#	void plot_start(int, int)		to	svgPlot; .plotStart
#	void plot_end(int)			to	svgPlot; .plotEnd
#	void move(double, double)		to	svgPlot; .move
#	void move_rel(double, double)		to	svgPlot; .moveRel
#	void draw(double, double)		to	svgPlot; .draw
#	void draw_rel(double, double)		to	svgPlot; .drawRel
#

import sys

class svgPlot:
    def __init__(self, x, y):
        self.fh = None
        self.Y = y
        self.header = '<svg xmlns="http://www.w3.org/2000/svg" version="1.1" width="{0}" height="{1}">'.format(x, y)
        self.footer = "</svg>"

    def plotStart(self, fileHandler=None):
        self.fh = fileHandler != None and fileHandler or sys.stdout

        self.fh.write(self.header)
        self.fh.write("\n")
        self.fh.write('<path d="')

    def plotEnd(self, isClosePath=False):
        if isClosePath:
            self.fh.write("Z")
        self.fh.write('" fill="none" stroke="black" />')
        self.fh.write("\n")
        self.fh.write(self.footer)
        self.fh.write("\n")

        self.fh = None

    def move(self, x, y):
        self.fh.write("M {0:g} {1:g} ".format(x, self.Y - y))

    def moveRel(self, x, y):
        self.fh.write("m {0:g} {1:g} ".format(x, -y))

    def draw(self, x, y):
        self.fh.write("L {0:g} {1:g} ".format(x, self.Y - y))

    def drawRel(self, x, y):
        self.fh.write("l {0:g} {1:g} ".format(x, -y))

File: algo-c-to-_/src/svgplot.lua
--
--	from src/svgplot.c
--
--	void plot_start(int, int)		to	svgPlot; :plotStart
--	void plot_end(int)			to	svgPlot; :plotEnd
--	void move(double, double)		to	svgPlot; :move
--	void move_rel(double, double)		to	svgPlot; :moveRel
--	void draw(double, double)		to	svgPlot; :draw
--	void draw_rel(double, double)		to	svgPlot; :drawRel
--

local function header(x, y)
	return ([[
<svg xmlns="http://www.w3.org/2000/svg" version="1.1" width="%d" height="%d">]]):format(x, y)
end

local function footer()
	return "</svg>"
end

local function pathStart()
	return [[<path d="]]
end

local function pathEnd(isClosePath)
	return ([[%s" fill="none" stroke="black" />]]):format(isClosePath and "Z" or "")
end

local function svgPlot(X, Y)
	local T = { fh = nil }

	function T:plotStart(fileHandler)
		T.fh = fileHandler ~= nil and fileHandler or io.stdout
		T.fh:write(header(X, Y), "\n")
		T.fh:write(pathStart())
	end

	function T:plotEnd(isClosePath)
		T.fh:write(pathEnd(isClosePath), "\n")
		T.fh:write(footer(), "\n")
		T.fh = nil
	end

	function T:move(x, y)
		T.fh:write(("M %g %g "):format(x, Y - y))
	end

	function T:moveRel(x, y)
		T.fh:write(("m %g %g "):format(x, -y))
	end

	function T:draw(x, y)
		T.fh:write(("L %g %g "):format(x, Y - y))
	end

	function T:drawRel(x, y)
		T.fh:write(("l %g %g "):format(x, -y))
	end

	return T
end

return {
	svgPlot = svgPlot
}
